### PR TITLE
Make SVI a TracePosterior subclass

### PR DESCRIPTION
**Pull request by eb8680**
Part of #1227 

This PR makes SVI a `TracePosterior` instance again, allowing it to be used with `EmpiricalMarginal` and more easily swapped out with `MCMC` and other inference algorithms.  It also updates `contrib.oed` to reflect this change.

### CODE REVIEW

**Comment by fritzo** [comment]
@eb8680 I think your implementation is correct. It looks like your test failures are due to new stricter plumbing that exposes latent bugs (probably my fault). Let's blame these errors on the dangerous practice of discarding unused `**kwargs`.

**Comment by ae-foster** [comment]
Looks good @eb8680 - should you write a test for this new way of running SVI (with `.run(...)`) or update any old SVI tests? From the OED side, things looks much cleaner.

File: pyro/infer/svi.py [hunk 1]
```diff
@@ -51,6 +59,19 @@ def _loss_and_grads(*args, **kwargs):
...
+            with poutine.block():
+                for i in range(self.num_steps):
+                    self.step(*args, **kwargs)
+        return super(SVI, self).run(*args, **kwargs)
+
+    def _traces(self, *args, **kwargs):
+        for i in range(self.num_samples):
+            guide_trace = poutine.trace(self.guide).get_trace(*args, **kwargs)
+            model_trace = poutine.trace(poutine.replay(self.model, trace=guide_trace)).get_trace(*args, **kwargs)
+            yield model_trace, (model_trace.log_prob_sum() - guide_trace.log_prob_sum())
```
Comment (neerajprad) [inline pyro/infer/svi.py]: We can probably set the importance weights to some constant here, right (seems like double adjustment otherwise)?

Comment (eb8680) [reply pyro/infer/svi.py]: Yeah, you're right

File: pyro/contrib/oed/eig.py [hunk 2]
```diff
@@ -45,17 +46,12 @@ def posterior_entropy(y_dist, design):
         y = pyro.sample("conditioning_y", y_dist)
         y_dict = {label: y[i, ...] for i, label in enumerate(observation_labels)}
         conditioned_model = pyro.condition(model, data=y_dict)
-        posterior = SVI(conditioned_model, vi_parameters["guide"],
-                        vi_parameters["optim"], loss=Trace_ELBO())
-        with pyro.poutine.block():
-            for _ in range(vi_parameters["num_steps"]):
-                posterior.step(design)
+        posterior = SVI(conditioned_model, **vi_parameters).run(design)  # noqa: F841
```
Comment (fritzo) [inline pyro/contrib/oed/eig.py]: nit: Why not remove the `posterior =` as suggested by flake8?

**Pull request merged**